### ISSUE TITLE
java.util.regex.PatternSyntaxException: Unclosed character class near index 5 .*/[ab

### ISSUE THREAD

**Issue by ppalaga** [angelozerr/ec4j#13]
Steps to reproduce: 

(1) Check out d9be2e04456b0ac5d68dc84b364753506b689347

(2) `mvn clean test`

Expected: all tests pass

Actual:

```
[ERROR] brackets_close_inside(org.eclipse.ec4j.EditorConfigGlobTest)  Time elapsed: 0.001 s  <<< ERROR!
org.eclipse.ec4j.EditorConfigException
        at org.eclipse.ec4j.AbstractEditorConfigManager.getOptions(AbstractEditorConfigManager.java:97)
        at org.eclipse.ec4j.EditorConfigGlobTest.brackets_close_inside(EditorConfigGlobTest.java:148)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
        at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
        at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
        at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
        at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
        at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
        at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
        at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
        at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
        at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
        at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
        at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
        at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
        at org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:369)
        at org.apache.maven.surefire.junit4.JUnit4Provider.executeWithRerun(JUnit4Provider.java:275)
        at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:239)
        at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:160)
        at org.apache.maven.surefire.booter.ForkedBooter.invokeProviderInSameClassLoader(ForkedBooter.java:373)
        at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:334)
        at org.apache.maven.surefire.booter.ForkedBooter.execute(ForkedBooter.java:119)
        at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:407)
Caused by: java.util.regex.PatternSyntaxException: Unclosed character class near index 5
.*/[ab
     ^
        at java.util.regex.Pattern.error(Pattern.java:1955)
        at java.util.regex.Pattern.clazz(Pattern.java:2548)
        at java.util.regex.Pattern.sequence(Pattern.java:2063)
        at java.util.regex.Pattern.expr(Pattern.java:1996)
        at java.util.regex.Pattern.compile(Pattern.java:1696)
        at java.util.regex.Pattern.<init>(Pattern.java:1351)
        at java.util.regex.Pattern.compile(Pattern.java:1028)
        at org.eclipse.ec4j.model.Section.match(Section.java:122)
        at org.eclipse.ec4j.AbstractEditorConfigManager.getOptions(AbstractEditorConfigManager.java:84)
        ... 26 more
```

**Comment by angelozerr** [comment]
@ppalaga I have seen that, I'm working to improve that

**Comment by angelozerr** [comment]
@ppalaga I have improved a lot the parser.

On Windows I have:

```
78% tests passed, 39 tests failed out of 180

Total Test time (real) =  19.83 sec

The following tests FAILED:
          9 - brackets_close_inside (Failed)
         11 - brackets_nclose_inside (Failed)
56 - braces_escaped_comma1 (Failed)
         59 - braces_escaped_brace1 (Failed)
         60 - braces_escaped_brace2 (Failed)
         61 - braces_escaped_brace3 (Failed)
         84 - braces_alpha_range1 (Failed)
         90 - utf_8_char (Failed)
         97 - star_star_over_separator7 (Failed)
         98 - star_star_over_separator8 (Failed)
         99 - star_star_over_separator9 (Failed)
        100 - star_star_over_separator10 (Failed)
        101 - star_star_over_separator11 (Failed)
        102 - star_star_over_separator12 (Failed)
        103 - star_star_over_separator13 (Failed)
        104 - star_star_over_separator14 (Failed)
        105 - star_star_over_separator15 (Failed)
        106 - star_star_over_separator16 (Failed)
        107 - star_star_over_separator17 (Failed)
        108 - star_star_over_separator18 (Failed)
        109 - star_star_over_separator19 (Failed)
        110 - star_star_over_separator20 (Failed)
        111 - star_star_over_separator21 (Failed)
        112 - star_star_over_separator22 (Failed)
        113 - star_star_over_separator23 (Failed)
        114 - star_star_over_separator24 (Failed)
        148 - escaped_semicolon_in_section (Failed)
        155 - escaped_octothorpe_in_section (Failed)
        158 - max_property_name (Failed)
        159 - max_property_value (Failed)
        161 - max_section_name_ignore (Failed)
        163 - parent_and_current_dir (Failed)
        164 - parent_dir_overload (Failed)
        165 - parent_dir_overload_repeat (Failed)
        166 - parent_dir_overload_repeat2 (Failed)
        168 - root_file_mixed_case (Failed)
        175 - windows_separator (Failed)
        178 - test_long_version_switch (Failed)
        179 - test_short_version_switch (Failed)
```

I have tested the editorconfig-core-java and here the result:
```
81% tests passed, 34 tests failed out of 180

Total Test time (real) =  18.63 sec

The following tests FAILED:
          9 - brackets_close_inside (Failed)
         11 - brackets_nclose_inside (Failed)
         56 - braces_escaped_comma1 (Failed)
         59 - braces_escaped_brace1 (Failed)
         60 - braces_escaped_brace2 (Failed)
         61 - braces_escaped_brace3 (Failed)
         84 - braces_alpha_range1 (Failed)
         90 - utf_8_char (Failed)
         97 - star_star_over_separator7 (Failed)
         98 - star_star_over_separator8 (Failed)
         99 - star_star_over_separator9 (Failed)
        100 - star_star_over_separator10 (Failed)
        101 - star_star_over_separator11 (Failed)
        102 - star_star_over_separator12 (Failed)
        103 - star_star_over_separator13 (Failed)
        104 - star_star_over_separator14 (Failed)
        105 - star_star_over_separator15 (Failed)
        106 - star_star_over_separator16 (Failed)
        107 - star_star_over_separator17 (Failed)
        108 - star_star_over_separator18 (Failed)
        109 - star_star_over_separator19 (Failed)
        110 - star_star_over_separator20 (Failed)
        111 - star_star_over_separator21 (Failed)
        112 - star_star_over_separator22 (Failed)
        113 - star_star_over_separator23 (Failed)
        114 - star_star_over_separator24 (Failed)
        148 - escaped_semicolon_in_section (Failed)
        155 - escaped_octothorpe_in_section (Failed)
        158 - max_property_name (Failed)
        159 - max_property_value (Failed)
        160 - max_section_name_ok (Failed)
        161 - max_section_name_ignore (Failed)
        171 - path_separator_backslash_in_cmd_line (Failed)
        175 - windows_separator (Failed)
```

I will try to continue to improve the parser , but I'm not sure I will be 100% compliant with ctests (perhaps in the future, but I would like to create a release at the end of the month) . My goal is to have 81% (instead of having 78%) like editorconfig-core-java. (Hope you will understand)

In other words, travis error should not block the build. What do you think about that?

**Comment by ppalaga** [comment]
I am happy about each improvement! 

BTW, on Linux, with https://github.com/angelozerr/ec4j/pull/18 applied (e51ce5882442d9725213427efe7a505d02e73c97), I can see 

```
93% tests passed, 13 tests failed out of 182

Total Test time (real) =  15.12 sec

The following tests FAILED:
         63 - braces_escaped_backslash2 (Failed)
         85 - braces_alpha_range1 (Failed)
        159 - max_property_name (Failed)
        160 - max_property_value (Failed)
        162 - max_section_name_ignore (Failed)
        164 - parent_and_current_dir (Failed)
        165 - parent_dir_overload (Failed)
        166 - parent_dir_overload_repeat (Failed)
        167 - parent_dir_overload_repeat2 (Failed)
        169 - root_file_mixed_case (Failed)
        178 - backslash_not_on_windows (Failed)
        180 - test_long_version_switch (Failed)
        181 - test_short_version_switch (Failed)
```

> In other words, travis error should not block the build. What do you think about that?

You mean not having 100% core tests passing should not block the first release? Of course, once we are better than the competitor, we can release :). 

But we should still make sure that each commit passes the unit tests. The objective of the present PR is to make that clearly visible.

**Comment by ppalaga** [comment]
Fixed as of 60e4fce15cc48b43444b1318b7fcad2888539ebd

**Comment by ppalaga** [comment]
BTW, `editorconfig-core-java` 's score on Linux is 100%. Just tested with their master e3e090545f44d20f5f228ef1068af4c9d7323a51

**Comment by angelozerr** [comment]
@ppalaga `editorconfig-core-java` doesn't support the last version of test. Submodule is linked to tests
https://github.com/editorconfig/editorconfig-core-test/tree/abb579e00f2deeede91cb485e53512efab9c6474

For instance they doesn't support max_property_name /max_property_value tests (that I have implemented).

**Comment by ppalaga** [comment]
> editorconfig-core-java doesn't support the last version of test

You are right, the core tests submodule in my workspace was quite old. This is what I see when I upgraded to the latest master 

```
97% tests passed, 5 tests failed out of 182

Total Test time (real) =  10.43 sec

The following tests FAILED:
         85 - braces_alpha_range1 (Failed)
        159 - max_property_name (Failed)
        160 - max_property_value (Failed)
        161 - max_section_name_ok (Failed)
        162 - max_section_name_ignore (Failed)
```
